proteuslib/property_models/tests/property_test_harness.py

```diff
@@ -0,0 +1,532 @@
+                    "that sets up the PropertyUnitTestHarness".format(obj_str=obj_str, blk=blk))
+
+    @pytest.fixture(scope='class')
+    def frame(self):
+        m = ConcreteModel()
+        self.configure_class(m)
+
+        m.fs = FlowsheetBlock(default={"dynamic": False})
+        m.fs.properties = self.prop_pack()
+        m.fs.stream = m.fs.properties.build_state_block(
+            [0], default=self.param_args)
+
+        for (v_str, ind), sf in self.set_default_scaling_dict.items():
+            m.fs.properties.set_default_scaling(v_str, sf, index=ind)
+
+        return m
+
+    @pytest.mark.unit
+    def test_components_phases(self, frame):
+        m = frame
+        # components
+        assert isinstance(m.fs.properties.component_list, Set)
+        assert len(m.fs.properties.component_list) == len(m.component_dict)
+        for c_str in m.component_dict:
+            assert c_str in m.fs.properties.component_list
+            c = getattr(m.fs.properties, c_str)
+            c_type = m.component_dict[c_str]
+            assert isinstance(c, c_type)
+
+        # phases
+        assert isinstance(m.fs.properties.phase_list, Set)
+        assert len(m.fs.properties.phase_list) == len(m.phase_dict)
+        for p_str in m.phase_dict:
+            assert p_str in m.fs.properties.phase_list
+            p = getattr(m.fs.properties, p_str)
+            p_type = m.phase_dict[p_str]
+            assert isinstance(p, p_type)
+
+    @pytest.mark.unit
+    def test_parameters(self, frame):
+        m = frame
+
+        self.check_block_obj(m.fs.properties, m.param_obj_dict)
+        self.check_block_obj_coverage(m.fs.properties, m.param_obj_dict)
+
+        # test that the parameter variables are fixed
+        for v in m.fs.properties.component_objects(Var):
+            assert v.is_fixed()
+
+    @pytest.mark.unit
+    def test_default_scaling(self, frame):
+        m = frame
+
+        assert isinstance(m.fs.properties.default_scaling_factor, dict)
+        for t, sf in m.fs.properties.default_scaling_factor.items():
+            assert t in m.default_scaling_factor_dict.keys()
+            assert m.default_scaling_factor_dict[t] == sf
+
+    @pytest.mark.unit
+    def test_state_variables(self, frame):
+        m = frame
+        # state variables
+        state_vars_dict = m.fs.stream[0].define_state_vars()
+        assert len(state_vars_dict) == len(m.state_vars_list)
+        for v_str in state_vars_dict:
+            assert v_str in m.state_vars_list
+
+    @pytest.mark.unit
+    def test_metadata(self, frame):
+        m = frame
+        assert hasattr(m.fs.properties, 'define_metadata')
+        metadata = m.fs.properties.get_metadata().properties
+
+        # check method naming convention
+        for v_str in metadata.keys():
+            if v_str in m.state_vars_list:
+                assert metadata[v_str]['method'] is None
+            else:
+                assert metadata[v_str]['method'] == '_' + v_str
```

Comment:
Is this test really necessary? Would it not be better to just test that all the expected properties can be constructed?

Reply:
The test mostly just ensures that the on demand methods follow the standard naming convention, which is very minor. Do you think we shouldn't enforce this convention? 

The test_build checks that all expected properties can be constructed but I wanted there to be some test directly for the metadata, rather than an indirect test like if the properties can be constructed. 

I think I'll add a test that checks that the metadata includes the variables and expressions in the developer provided testing dictionary.

Reply:
The main problem here is that there is no documented naming convention for constraints. Making this a standard is not necessarily a problem, but it would need to be documented and conveyed to the developers.

I am not sure how much value there is to testing the metadata directly either - the metadata doesn't do a lot all up and the only important part is that build-on-demand works. You are also assuming that only state vars will have `"mehtod"=None` (and even assuming that as well).